src/odoc/fs.ml

```diff
@@ -39,7 +39,43 @@ module File = struct
     | Error (`Msg e) -> invalid_arg ("Odoc.Fs.File.of_string: " ^ e)
     | Ok p -> p
 
-  let read = Bos.OS.File.read
+  let read file =
+    let with_ic ~close ic f =
+      let close ic = try close ic with Sys_error _ -> () in
+      match f ic with v -> close ic; v | exception e -> close ic; raise e
+    in
+    let input_one_shot len ic =
+      let buf = Bytes.create len in
+      really_input ic buf 0 len;
+      close_in ic;
+      Ok (Bytes.unsafe_to_string buf)
+    in
+    let input_stream ic =
+      let bsize = 65536 (* IO_BUFFER_SIZE *) in
+      let buf = Buffer.create bsize in
+      let b = Bytes.create bsize in
+      let rec loop () =
```

Comment:
Would [`Buffer.add_channel`](https://caml.inria.fr/pub/docs/manual-ocaml/libref/Buffer.html#VALadd_channel) work here? Not that we necessarily need this to be more copy-efficient, but nonetheless. This may also do away with the need for the distinction between `input_one_shot` and `input_stream`, assuming some exception is raised if the `Buffer.t` gets too large.

Reply:
> Would [`Buffer.add_channel`](https://caml.inria.fr/pub/docs/manual-ocaml/libref/Buffer.html#VALadd_channel) work here?

Yes. 

> This may also do away with the need for the distinction between `input_one_shot` and `input_stream`

Kept one shot. This is really the most efficient you can get to allocation wise: the exact result.